### PR TITLE
8270459: Conflict inlining decisions by C1/C2 with the same CompileCommand

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

Conflict inlining decisions can be observed by C1/C2 with the same CompileCommand.
And I came across this situation while I was analyzing a performance issue using the jmh, which made me quite confused.

For the example in the JBS.
```
     28    1 %     3       TestInline::main @ 11 (43 bytes)
                              @ 23   TestInline::caller (4 bytes)   disallowed by CompileCommand
     29    2       3       TestInline::main (43 bytes)
                              @ 23   TestInline::caller (4 bytes)   disallowed by CompileCommand
     31    3 %     4       TestInline::main @ 11 (43 bytes)
     33    1 %     3       TestInline::main @ 11 (43 bytes)   made not entrant
                              @ 23   TestInline::caller (4 bytes)   force inline by CompileCommand
```

This can be easily reproduced by specifying both the `inline` and `dontinline` compile commands for a method.
```
   -XX:CompileCommand=inline,TestInline::caller \
   -XX:CompileCommand=dontinline,TestInline::caller \
```

The fix will ignore the inline/not-inline directives if both the `inline` and `dontinline` compile commands are specified for a method.

Effect of the patch:
```
     27    1 %     3       TestInline::main @ 11 (43 bytes)
                              @ 23   TestInline::caller (4 bytes)   inline
     28    2       3       TestInline::main (43 bytes)
                              @ 23   TestInline::caller (4 bytes)   inline
     28    3 %     4       TestInline::main @ 11 (43 bytes)
     30    1 %     3       TestInline::main @ 11 (43 bytes)   made not entrant
                              @ 23   TestInline::caller (4 bytes)   inline (hot)
     31    3 %     4       TestInline::main @ 11 (43 bytes)   made not entrant
```

Testing:
  - tier 1~3 on Linux/x64, no regression

Thanks.
Best regards,
Jie

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270459](https://bugs.openjdk.java.net/browse/JDK-8270459): Conflict inlining decisions by C1/C2 with the same CompileCommand

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4780/head:pull/4780` \
`$ git checkout pull/4780`

Update a local copy of the PR: \
`$ git checkout pull/4780` \
`$ git pull https://git.openjdk.java.net/jdk pull/4780/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4780`

View PR using the GUI difftool: \
`$ git pr show -t 4780`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4780.diff">https://git.openjdk.java.net/jdk/pull/4780.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
In HotSpot the last specified option overwrites all previous.
I am not sure we keep information about CompileCommand options order but we have to follow general rule.

**Comment by DamonFool** [comment]
> In HotSpot the last specified option overwrites all previous.
> I am not sure we keep information about CompileCommand options order but we have to follow general rule.

Okay.
Fixed.
Thanks.

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (vnkozlov) [inline src/hotspot/share/compiler/compilerOracle.cpp:355]: Add comment that `option_list` lists options in reverse order. So the first option we find is the last which was specified.

Comment (DamonFool) [reply src/hotspot/share/compiler/compilerOracle.cpp:356]: > Add comment that `option_list` lists options in reverse order. So the first option we find is the last which was specified.

Updated.
Thanks.

**Review by vnkozlov — APPROVED** [review]

**Comment by vnkozlov** [comment]
You need second review.

File: src/hotspot/share/compiler/compilerOracle.cpp
Comment (navyxliu) [reply src/hotspot/share/compiler/compilerOracle.cpp:356]: if has_inline == true && has_dnotinline == true, but v1 == false && v2 == false,  two options conflict, don't  they? 
in this case, you go to else, which actually bias to Inline instead of the last seen option.

Comment (navyxliu) [reply src/hotspot/share/compiler/compilerOracle.cpp:356]: maybe like this:

Comment (DamonFool) [reply src/hotspot/share/compiler/compilerOracle.cpp:356]: > if has_inline == true && has_dnotinline == true, but v1 == false && v2 == false, two options conflict, don't they?

I don't think they are conflict.

This is because
```
-XX:CompilerCommand=inline,foo,false     ==> doesn't mean to disallow the inline of foo
-XX:CompilerCommand=dontinline,foo,false ==>  doesn't mean to force inline of foo
```

So if `inline` and `dontinline` options are set to false, they actually have no effect to the compiler.
Thanks.

Comment (navyxliu) [reply src/hotspot/share/compiler/compilerOracle.cpp:356]: make sense. I take back my comment.  those 2 options don't have any effect when they are false.

**Review by navyxliu — APPROVED** [review]

**Comment by DamonFool** [comment]
Thanks @vnkozlov and @navyxliu for your review.
/integrate

**Pull request closed without merging**
